### PR TITLE
🏗Refactor oidc Authentication

### LINKED ISSUE #702: Title Bar icons overlap with Identity Server Header
#### Describe your issue
The Icons of the title bar are in conflict with the Logout Page of the Identity Server. 

#### How can others reproduce the issue?
* Log in
* Log out
* Look at the title bar of the opened Identity Server Window

screen shot 2018-08-02 at 13 37 41

#### Possible solution
Add a solid Task bar

#### Issue checklist

Please check the boxes in this list after submitting your Issue:

- [x] I've checked if this issue already exists
- [x] I've included all the information that i think is relevant
- [x] I've added logs and/or screenshots (if applicable)
- [x] I've mentioned PRs and issues that relate to this one

### LINKED ISSUE #705: Electron@2/3 not working
#### Describe your issue

Currently we're having compatibility issues with Electron@2.

This issue aims to track features of Electron@2 we want to use and problems that prevent us from upgrading to Electron@2.

#### New Features we want to use

* File Type Filter in Open File/Folder Dialog
* Dedicated Title Bar
   * No Overlap between Window Buttons and e.g. IdentityServer Page Header

#### Known Problems

* Redirect from IdentityServer doesn't work, results in empty window

#### Issue checklist

Please check the boxes in this list after submitting your Issue:

- [x] I've checked if this issue already exists
- [x] I've included all the information that i think is relevant
- [x] I've added logs and/or screenshots (if applicable)
- [x] I've mentioned PRs and issues that relate to this one

### PR DESCRIPTION
**Pull request by Vyperus**
**Changes:**

This PR Introduces big changes to the overall behavior on login with identity server.
It is the first step to refactor the authentication mechanism in BPMN-Studio and allows us to use Electron 4 from now on. 

1. Replace old login/logout flow with a new one, suitable for Electron 4.
2. Update Electron to v4.0.1.
3. Fixes necessary Parts for new Electron version.
4. Add Credentials to every remote Solution. (Subject to change)

**Issues:**

Closes #702 
Closes #705

PR: #1295

## How can others test the changes?

1. Reinstall all node modules.
3. Get the identity server: `https://github.com/process-engine/identity_server`.
3. Start the identity server with `dotnet restore` and `dotnet run`.
4. Start BPMN-Studio.
5. Click the Login Button. (A second window should open now)
6. Login
7. See all requests against remote solutions have a real access_token now.
8. Logout.
9. See that all requests use the dummy token again.

## PR-Checklist

Please check the boxes in this list after submitting your PR:

- [x] You can merge this PR **right now** (if not, please prefix the title with "WIP: ")
- [x] I've tested **all** changes included in this PR.
- [x] I've also reviewed this PR myself before submitting (e.g. for scrambled letters, typos, etc.).
- [x] I've rebased the `develop` branch with my branch before finishing this PR.
- [x] I've **summarized all changes** in a list above.
- [x] I've mentioned all **PRs, which relate to this one**.
- [x] I've prefixed my Pull Request title is according to [gitmoji guide](https://gitmoji.carloscuesta.me/).

### CODE REVIEW

File: src/modules/authentication/electron.oidc.authentication.service.ts [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+const IDENTITY_SERVER_AVAILABLE_SUCCESS_STATUS_CODE: number = 200;
+
+@inject(EventAggregator, 'NotificationService', Router, 'SolutionService')
+export class ElectronOidcAuthenticationService implements IAuthenticationService {
+
+  private _eventAggregator: EventAggregator;
+  private _notificationService: NotificationService;
+  private _router: Router;
+  private _tokenObject: ITokenObject;
+  private _logoutWindow: Window = null;
```
[the remark this answers is not in the document]
Comment (Vyperus) [reply src/modules/authentication/electron.oidc.authentication.service.ts]: No, that is just a leftover.

File: electron_app/electron.js [hunk 2]
```diff
@@ -6,6 +6,9 @@ const app = electron.app;
 const isDev = require('electron-is-dev');
 const getPort = require('get-port');
 const fs = require('fs');
+const electronOidc = require('./electron-oidc');
+const oidcConfig = require('./oidc-config');
+const ipcMain = electron.ipcMain;
```
[the remark this answers is not in the document]
Comment (cmg-dev) [reply electron_app/electron.js]: The deconstruct syntax is always favourable.  Please use @LeoTT suggestion.

File: src/modules/authentication/electron.oidc.authentication.service.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (cmg-dev) [reply src/modules/authentication/electron.oidc.authentication.service.ts]: so delete it, please.

File: src/modules/authentication/electron.oidc.authentication.service.ts [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+      this._notificationService.showNotification(NotificationType.WARNING, 'The identity server may be offline!');
+      this._tokenObject = undefined;
+
+      this._logoutUserFromAllSolutions();
+    }
+
+  }
+
+  private async _isIdentityServerReachable(): Promise<boolean> {
+    const request: Request = new Request(`${environment.openIdConnect.authority}/.well-known/openid-configuration`, {
```
Comment (cmg-dev) [inline src/modules/authentication/electron.oidc.authentication.service.ts]: same.

File: src/modules/authentication/index.ts [hunk 4]
```diff
@@ -1,6 +1,14 @@
 import {FrameworkConfiguration} from 'aurelia-framework';
-import {AuthenticationService} from './authentication.service';
+import {ElectronOidcAuthenticationService} from './electron.oidc.authentication.service';
+import {WebOidcAuthenticationService} from './web.oidc.authentication.service';
 
 export async function configure(config: FrameworkConfiguration): Promise<void> {
-  config.container.registerSingleton('AuthenticationService', AuthenticationService);
+  const appIsRunningInElectron: boolean = (window as any).nodeRequire;
+
```
Comment (cmg-dev) [inline src/modules/authentication/index.ts]: rm newline.

File: src/modules/authentication/electron.oidc.authentication.service.ts [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+  // enables us to work without depending on a full environment with
+  // IdentityServer.
+  private _getDummyAccessToken(): string {
+    const dummyAccessTokenString: string = 'dummy_token';
+    const base64EncodedString: string = btoa(dummyAccessTokenString);
+    return base64EncodedString;
+  }
+
+  private _persistTokenObject(): void {
+    window.localStorage.setItem('tokenObject', JSON.stringify(this._tokenObject));
```
Comment (cmg-dev) [inline src/modules/authentication/electron.oidc.authentication.service.ts]: Probably we need to have a namespace like thing:

```suggestion
    window.localStorage.setItem('bpmnstudio.tokenObject', JSON.stringify(this._tokenObject));
```

I'm afraid, we might collide with other applications.

File: electron_app/electron-oidc.js [hunk 6]
```diff
@@ -0,0 +1,168 @@
+'use strict';
+
+const Promise = require('pinkie-promise');
+const queryString = require('querystring');
+const fetch = require('node-fetch');
```
[the remark this answers is not in the document]
Comment (Vyperus) [reply electron_app/electron-oidc.js]: We are technically not in the browser context here but in the electron context where this works fine.

File: electron_app/electron.js [hunk 7]
```diff
@@ -223,8 +148,43 @@ Main._initializeApplication = function () {
...
+          console.log('Error while getting token', err);
+        });
+    });
+
+    ipcMain.on('oidc-logout', (event, tokenObject, authorityUrl) => {
+      electronOidcInstance.logout(tokenObject, authorityUrl)
+        .then(logoutWasSuccessful => {
+          event.sender.send('oidc-logout-reply', logoutWasSuccessful);
+        }, err => {
+          console.log('Error while logging out', err);
```
[the remark this answers is not in the document]
Comment (Vyperus) [reply electron_app/electron.js]: I can send the event reply with an error, check it in the service and display a message there.

**Pull request merged**